### ISSUE TITLE
"(psycopg2.ProgrammingError) can't adapt type 'WKBElement'" when updating to 0.8

### ISSUE THREAD

**Issue by FrankyBoy** [geoalchemy/geoalchemy2#291]
Hi!

This code works in 0.7.0, but not any more in 0.8.4. SQLAlchemy is at 1.3.18.

```
[definition ...]

class Object: 
    [...other properties ... ]
    geom = db.Column(Geometry(geometry_type='GEOMETRY', srid=3857, dimension=2, spatial_index=True, management=False))

[ ... in usage ... ] 
from sqlalchemy import func
obj = Object.query.filter_by(id=obj_id).first()
geom_type = db.session.execute(func.GeometryType(obj.geom)).first().values()[0]
```

The error raised is "(psycopg2.ProgrammingError) can't adapt type 'WKBElement'". 

Let me know if you need additional infos. 

E: add the import to clarify where "func" comes from.

**Comment by adrien-berchet** [comment]
Hi @FrankyBoy 
Thanks for this report.
I understand the issue, but the fix might not be trivial. It is due to a combination of things:
* `WKBElement` is no more a subclass of `sqlalchemy.sql.functions.Function`, so it can not be automatically processed by `psycopg2`.
* `GeometryType` is not declared in `geoalchemy2.functions`.
* `GeometryType` does not start with the `ST_` prefix.
I will think about it...

For now, if you can use the function `ST_GeometryType` it should work. Otherwise you can define the function yourself (see example here: https://geoalchemy-2.readthedocs.io/en/latest/spatial_functions.html#geoalchemy2.functions.GenericFunction) but it will not work with the ORM (you will not be able to write something like `obj.geom.GeometryType()`) as it does not start with `ST_*`.

**Comment by FrankyBoy** [comment]
While I appreciate this answer, it raises two problems for me: 

1. the results of ST_GeometryType and GeometryType are not the same, meaning a great deal of effort in analysis of the usages and the possible impact. _Maybe_ the result of ST_GeometryType can be changed to resemble GeometryType (i.e. substring plus uppercase), though the postgis documentation is not very clear on this. 
2. It leaves open the general question of how many more functions have stopped working due to this change, which means a great deal of regression testing and investigation on possible impact across the whole code base for something that should be a minor version update of a library. The changelog does not contain a list of functions affected so for me this is a block in upgrading simply due to the efforts involved.

**Comment by adrien-berchet** [comment]
I agree on both of your points. For the first one, it was just a possible workaround. For the second one, most of the functions have been added to `geoalchemy.functions` to reduce the possible impacts of this regression (which should have been notified in the release notes, sorry about that). The reasons for this regression are explained in #248 and thus the `_SpatialElement` (base class for `WKBElement`, `WKTElement` and `RasterElement`) can no longer be SQL expressions.

I see 2 ways to fix this:
1. add all functions, even if they do not start with the `ST_` prefix and update the `__getattr__()` method of the `_SpatialElement` class to replace the test `if not name.lower().startswith('st_')` by something like `if name.lower() not in the_list_of_all_function_names`.
2. register an adapter for the `_SpatialElement` class in `psycopg2` but I never tried this with `SQLite` (though it seems to work in the same way: https://docs.python.org/3.6/library/sqlite3.html#sqlite3.register_adapter).

The option 2 looks easier to maintain since it will work for any function that is defined in the DB (it could even work for custom functions). But the option 1 is needed anyway for the ORM to work with functions without the `ST_` prefix. So maybe we should do both...

WDYT @elemoine?

**Comment by adrien-berchet** [comment]
Hi @FrankyBoy 
The last release should support function without the `ST_` suffix. The function `GeometryType` is still not defined in `geoalchemy2` but now you can define it on your side (https://geoalchemy-2.readthedocs.io/en/latest/spatial_functions.html#geoalchemy2.functions.GenericFunction).
The next step would be to ensure that all the Postgis functions are registered in `geoalchemy2` but it is not for tomorrow :-) (though an attempt is on the way in #298 )

**Comment by adrien-berchet** [comment]
Should be fixed now that #298 is merged
